File: backend/app/collectors/mock.py
"""Mock collectors that generate fake data for testing."""

import random
from datetime import datetime

from app.collectors.base import BaseCollector
from app.models.metrics import (
    ContainerMetrics,
    CpuMetrics,
    DiskMetrics,
    DockerMetrics,
    LinuxMetrics,
    MemoryMetrics,
    NetworkMetrics,
    QBittorrentMetrics,
    SystemMetrics,
    SystemStatus,
    SystemType,
    TemperatureMetrics,
    TorrentInfo,
)


class MockLinuxCollector(BaseCollector):
    """Mock Linux collector with realistic random data."""

    def __init__(self, system_id: str, name: str, config: dict):
        super().__init__(system_id, name, SystemType.LINUX)
        self.config = config
        self._base_cpu = random.uniform(10, 40)
        self._base_mem = random.uniform(30, 60)
        self._uptime = random.randint(86400, 864000)

    async def check_connection(self) -> bool:
        return True

    def determine_status(self, metrics: LinuxMetrics) -> SystemStatus:
        if metrics.cpu.percent > 90 or metrics.memory.percent > 95:
            return SystemStatus.CRITICAL
        if metrics.cpu.percent > 75 or metrics.memory.percent > 85:
            return SystemStatus.WARNING
        return SystemStatus.HEALTHY

    async def collect(self) -> SystemMetrics:
        cpu_percent = max(0, min(100, self._base_cpu + random.uniform(-10, 15)))
        mem_percent = max(0, min(100, self._base_mem + random.uniform(-5, 10)))

        total_mem = 16 * 1024 * 1024 * 1024  # 16GB
        used_mem = int(total_mem * mem_percent / 100)

        core_count = 8
        core_percents = [max(0, min(100, cpu_percent + random.uniform(-20, 20))) for _ in range(core_count)]

        linux_metrics = LinuxMetrics(
            cpu=CpuMetrics(
                percent=cpu_percent,
                cores=core_percents,
                load_avg=[cpu_percent / 25, cpu_percent / 30, cpu_percent / 35],
            ),
            memory=MemoryMetrics(
                total=total_mem,
                used=used_mem,
                available=total_mem - used_mem,
                percent=mem_percent,
            ),
            disks=[
                DiskMetrics(
                    mount="/",
                    total=500 * 1024 * 1024 * 1024,
                    used=int(500 * 1024 * 1024 * 1024 * 0.45),
                    free=int(500 * 1024 * 1024 * 1024 * 0.55),
                    percent=45.0 + random.uniform(-2, 2),
                ),
                DiskMetrics(
                    mount="/home",
                    total=1000 * 1024 * 1024 * 1024,
                    used=int(1000 * 1024 * 1024 * 1024 * 0.62),
                    free=int(1000 * 1024 * 1024 * 1024 * 0.38),
                    percent=62.0 + random.uniform(-2, 2),
                ),
            ],
            network=NetworkMetrics(
                bytes_sent=random.randint(1000000000, 5000000000),
                bytes_recv=random.randint(5000000000, 20000000000),
                upload_rate=random.uniform(100000, 5000000),
                download_rate=random.uniform(500000, 10000000),
            ),
            temperatures=[
                TemperatureMetrics(label="CPU", current=45 + random.uniform(-5, 15)),
                TemperatureMetrics(label="GPU", current=40 + random.uniform(-5, 10)),
            ],
            uptime=self._uptime,
        )

        self._uptime += 5

        return SystemMetrics(
            id=self.system_id,
            name=self.name,
            type=SystemType.LINUX,
            status=self.determine_status(linux_metrics),
            last_updated=datetime.utcnow(),
            metrics=linux_metrics,
        )


class MockDockerCollector(BaseCollector):
    """Mock Docker collector."""

    def __init__(self, system_id: str, name: str, config: dict):
        super().__init__(system_id, name, SystemType.DOCKER)
        self.config = config
        self._containers = [
            ("nginx", "running"),
            ("postgres", "running"),
            ("redis", "running"),
            ("api-server", "running"),
            ("worker", "running"),
            ("monitoring", "stopped"),
        ]

    async def check_connection(self) -> bool:
        return True

    async def collect(self) -> SystemMetrics:
        containers = []
        running = 0
        stopped = 0

        for name, state in self._containers:
            is_running = state == "running"
            if is_running:
                running += 1
            else:
                stopped += 1

            containers.append(
                ContainerMetrics(
                    id=f"mock_{name}",
                    name=name,
                    image=f"{name}:latest",
                    state=state,
                    status="Up 3 days" if is_running else "Exited (0) 2 hours ago",
                    cpu_percent=random.uniform(0.5, 15) if is_running else 0,
                    memory_percent=random.uniform(1, 25) if is_running else 0,
                    memory_usage=random.randint(50000000, 500000000) if is_running else 0,
                    memory_limit=1024 * 1024 * 1024,
                )
            )

        docker_metrics = DockerMetrics(
            containers=containers,
            total_count=len(self._containers),
            running_count=running,
            stopped_count=stopped,
        )

        status = SystemStatus.HEALTHY
        if stopped > running:
            status = SystemStatus.WARNING

        return SystemMetrics(
            id=self.system_id,
            name=self.name,
            type=SystemType.DOCKER,
            status=status,
            last_updated=datetime.utcnow(),
            metrics=docker_metrics,
        )


class MockQBittorrentCollector(BaseCollector):
    """Mock qBittorrent collector."""

    def __init__(self, system_id: str, name: str, config: dict):
        super().__init__(system_id, name, SystemType.QBITTORRENT)
        self.config = config
        self._torrents = [
            {"name": "Ubuntu 24.04 LTS", "size": 4.5 * 1024**3, "progress": 100, "state": "stalledUP"},
            {"name": "Debian 12.0", "size": 3.8 * 1024**3, "progress": 100, "state": "uploading"},
            {"name": "Fedora 40", "size": 2.1 * 1024**3, "progress": 78.5, "state": "downloading"},
            {"name": "Arch Linux", "size": 850 * 1024**2, "progress": 45.2, "state": "downloading"},
            {"name": "Linux Mint 21.3", "size": 2.8 * 1024**3, "progress": 92.1, "state": "downloading"},
        ]

    async def check_connection(self) -> bool:
        return True

    async def collect(self) -> SystemMetrics:
        torrents = []
        active_downloads = 0
        total_download_speed = 0
        total_upload_speed = 0

        for t in self._torrents:
            # Simulate progress
            if t["progress"] < 100 and t["state"] == "downloading":
                t["progress"] = min(100, t["progress"] + random.uniform(0.1, 0.5))
                if t["progress"] >= 100:
                    t["state"] = "stalledUP"
                    t["progress"] = 100

            is_downloading = t["state"] in ("downloading", "forcedDL")
            is_uploading = t["state"] in ("uploading", "stalledUP", "forcedUP")

            dl_speed = int(random.uniform(1, 10) * 1024 * 1024) if is_downloading else 0
            ul_speed = int(random.uniform(0.1, 2) * 1024 * 1024) if is_uploading else 0

            if is_downloading:
                active_downloads += 1
                total_download_speed += dl_speed

            total_upload_speed += ul_speed

            eta = None
            if is_downloading and dl_speed > 0:
                remaining = t["size"] * (100 - t["progress"]) / 100
                eta = int(remaining / dl_speed)

            torrents.append(
                TorrentInfo(
                    name=t["name"],
                    size=int(t["size"]),
                    progress=t["progress"],
                    state=t["state"],
                    download_speed=dl_speed,
                    upload_speed=ul_speed,
                    eta=eta,
                )
            )

        qbit_metrics = QBittorrentMetrics(
            download_speed=total_download_speed,
            upload_speed=total_upload_speed,
            active_downloads=active_downloads,
            total_torrents=len(torrents),
            torrents=torrents,
        )

        return SystemMetrics(
            id=self.system_id,
            name=self.name,
            type=SystemType.QBITTORRENT,
            status=SystemStatus.HEALTHY,
            last_updated=datetime.utcnow(),
            metrics=qbit_metrics,
        )


MOCK_COLLECTORS = {
    "linux": MockLinuxCollector,
    "docker": MockDockerCollector,
    "qbittorrent": MockQBittorrentCollector,
}

File: backend/src/collectors/mock.rs
use async_trait::async_trait;
use chrono::Utc;
use rand::Rng;

use super::Collector;
use crate::models::*;

// --- Mock Linux ---

pub struct MockLinuxCollector {
    system_id: String,
    name: String,
    base_cpu: f64,
    base_mem: f64,
    uptime: i64,
}

impl MockLinuxCollector {
    pub fn new(system_id: String, name: String) -> Self {
        let mut rng = rand::thread_rng();
        Self {
            system_id,
            name,
            base_cpu: rng.gen_range(10.0..40.0),
            base_mem: rng.gen_range(30.0..60.0),
            uptime: rng.gen_range(86400..864000),
        }
    }
}

#[async_trait]
impl Collector for MockLinuxCollector {
    async fn collect(&mut self) -> SystemMetrics {
        let mut rng = rand::thread_rng();
        let cpu_percent = (self.base_cpu + rng.gen_range(-10.0..15.0)).clamp(0.0, 100.0);
        let mem_percent = (self.base_mem + rng.gen_range(-5.0..10.0)).clamp(0.0, 100.0);

        let total_mem: i64 = 16 * 1024 * 1024 * 1024;
        let used_mem = (total_mem as f64 * mem_percent / 100.0) as i64;

        let core_percents: Vec<f64> = (0..8)
            .map(|_| (cpu_percent + rng.gen_range(-20.0..20.0)).clamp(0.0, 100.0))
            .collect();

        let linux_metrics = LinuxMetrics {
            cpu: CpuMetrics {
                percent: cpu_percent,
                cores: core_percents,
                load_avg: vec![
                    cpu_percent / 25.0,
                    cpu_percent / 30.0,
                    cpu_percent / 35.0,
                ],
            },
            memory: MemoryMetrics {
                total: total_mem,
                used: used_mem,
                available: total_mem - used_mem,
                percent: mem_percent,
            },
            disks: vec![
                DiskMetrics {
                    mount: "/".to_string(),
                    total: 500 * 1024 * 1024 * 1024,
                    used: (500.0 * 1024.0 * 1024.0 * 1024.0 * 0.45) as i64,
                    free: (500.0 * 1024.0 * 1024.0 * 1024.0 * 0.55) as i64,
                    percent: 45.0 + rng.gen_range(-2.0..2.0),
                },
                DiskMetrics {
                    mount: "/home".to_string(),
                    total: 1000 * 1024 * 1024 * 1024,
                    used: (1000.0 * 1024.0 * 1024.0 * 1024.0 * 0.62) as i64,
                    free: (1000.0 * 1024.0 * 1024.0 * 1024.0 * 0.38) as i64,
                    percent: 62.0 + rng.gen_range(-2.0..2.0),
                },
            ],
            network: NetworkMetrics {
                bytes_sent: rng.gen_range(1_000_000_000..5_000_000_000),
                bytes_recv: rng.gen_range(5_000_000_000..20_000_000_000),
                upload_rate: rng.gen_range(100_000.0..5_000_000.0),
                download_rate: rng.gen_range(500_000.0..10_000_000.0),
            },
            temperatures: vec![
                TemperatureMetrics {
                    label: "CPU".to_string(),
                    current: 45.0 + rng.gen_range(-5.0..15.0),
                    high: None,
                    critical: None,
                },
                TemperatureMetrics {
                    label: "GPU".to_string(),
                    current: 40.0 + rng.gen_range(-5.0..10.0),
                    high: None,
                    critical: None,
                },
            ],
            uptime: self.uptime,
        };

        self.uptime += 5;

        let status = if linux_metrics.cpu.percent > 90.0 || linux_metrics.memory.percent > 95.0 {
            SystemStatus::Critical
        } else if linux_metrics.cpu.percent > 75.0 || linux_metrics.memory.percent > 85.0 {
            SystemStatus::Warning
        } else {
            SystemStatus::Healthy
        };

        SystemMetrics {
            id: self.system_id.clone(),
            name: self.name.clone(),
            system_type: SystemType::Linux,
            status,
            last_updated: Utc::now(),
            error: None,
            metrics: MetricsPayload::Linux(linux_metrics),
        }
    }

    async fn close(&mut self) {}
}

// --- Mock Docker ---

pub struct MockDockerCollector {
    system_id: String,
    name: String,
    containers: Vec<(&'static str, &'static str)>,
}

impl MockDockerCollector {
    pub fn new(system_id: String, name: String) -> Self {
        Self {
            system_id,
            name,
            containers: vec![
                ("nginx", "running"),
                ("postgres", "running"),
                ("redis", "running"),
                ("api-server", "running"),
                ("worker", "running"),
                ("monitoring", "stopped"),
            ],
        }
    }
}

#[async_trait]
impl Collector for MockDockerCollector {
    async fn collect(&mut self) -> SystemMetrics {
        let mut rng = rand::thread_rng();
        let mut container_metrics = Vec::new();
        let mut running = 0i32;
        let mut stopped = 0i32;

        for (name, state) in &self.containers {
            let is_running = *state == "running";
            if is_running {
                running += 1;
            } else {
                stopped += 1;
            }

            container_metrics.push(ContainerMetrics {
                id: format!("mock_{name}"),
                name: name.to_string(),
                image: format!("{name}:latest"),
                state: state.to_string(),
                status: if is_running {
                    "Up 3 days".to_string()
                } else {
                    "Exited (0) 2 hours ago".to_string()
                },
                cpu_percent: if is_running {
                    rng.gen_range(0.5..15.0)
                } else {
                    0.0
                },
                memory_percent: if is_running {
                    rng.gen_range(1.0..25.0)
                } else {
                    0.0
                },
                memory_usage: if is_running {
                    rng.gen_range(50_000_000..500_000_000)
                } else {
                    0
                },
                memory_limit: 1024 * 1024 * 1024,
            });
        }

        let status = if stopped > running {
            SystemStatus::Warning
        } else {
            SystemStatus::Healthy
        };

        SystemMetrics {
            id: self.system_id.clone(),
            name: self.name.clone(),
            system_type: SystemType::Docker,
            status,
            last_updated: Utc::now(),
            error: None,
            metrics: MetricsPayload::Docker(DockerMetrics {
                containers: container_metrics,
                total_count: self.containers.len() as i32,
                running_count: running,
                stopped_count: stopped,
            }),
        }
    }

    async fn close(&mut self) {}
}

// --- Mock qBittorrent ---

pub struct MockQBittorrentCollector {
    system_id: String,
    name: String,
    torrents: Vec<MockTorrent>,
}

struct MockTorrent {
    name: &'static str,
    size: i64,
    progress: f64,
    state: String,
}

impl MockQBittorrentCollector {
    pub fn new(system_id: String, name: String) -> Self {
        Self {
            system_id,
            name,
            torrents: vec![
                MockTorrent {
                    name: "Ubuntu 24.04 LTS",
                    size: (4.5 * 1024.0 * 1024.0 * 1024.0) as i64,
                    progress: 100.0,
                    state: "stalledUP".to_string(),
                },
                MockTorrent {
                    name: "Debian 12.0",
                    size: (3.8 * 1024.0 * 1024.0 * 1024.0) as i64,
                    progress: 100.0,
                    state: "uploading".to_string(),
                },
                MockTorrent {
                    name: "Fedora 40",
                    size: (2.1 * 1024.0 * 1024.0 * 1024.0) as i64,
                    progress: 78.5,
                    state: "downloading".to_string(),
                },
                MockTorrent {
                    name: "Arch Linux",
                    size: (850.0 * 1024.0 * 1024.0) as i64,
                    progress: 45.2,
                    state: "downloading".to_string(),
                },
                MockTorrent {
                    name: "Linux Mint 21.3",
                    size: (2.8 * 1024.0 * 1024.0 * 1024.0) as i64,
                    progress: 92.1,
                    state: "downloading".to_string(),
                },
            ],
        }
    }
}

#[async_trait]
impl Collector for MockQBittorrentCollector {
    async fn collect(&mut self) -> SystemMetrics {
        let mut rng = rand::thread_rng();
        let mut torrent_infos = Vec::new();
        let mut active_downloads = 0i32;
        let mut total_download_speed: i64 = 0;
        let mut total_upload_speed: i64 = 0;

        for (idx, t) in self.torrents.iter_mut().enumerate() {
            if t.progress < 100.0 && t.state == "downloading" {
                t.progress = (t.progress + rng.gen_range(0.1..0.5)).min(100.0);
                if t.progress >= 100.0 {
                    t.state = "stalledUP".to_string();
                    t.progress = 100.0;
                }
            }

            let is_downloading = t.state == "downloading" || t.state == "forcedDL";
            let is_uploading =
                t.state == "uploading" || t.state == "stalledUP" || t.state == "forcedUP";

            let dl_speed = if is_downloading {
                (rng.gen_range(1.0..10.0) * 1024.0 * 1024.0) as i64
            } else {
                0
            };
            let ul_speed = if is_uploading {
                (rng.gen_range(0.1..2.0) * 1024.0 * 1024.0) as i64
            } else {
                0
            };

            if is_downloading {
                active_downloads += 1;
                total_download_speed += dl_speed;
            }
            total_upload_speed += ul_speed;

            let eta = if is_downloading && dl_speed > 0 {
                let remaining = (t.size as f64 * (100.0 - t.progress) / 100.0) as i64;
                Some(remaining / dl_speed)
            } else {
                None
            };

            torrent_infos.push(TorrentInfo {
                hash: format!("{:032x}", t.name.len() as u128 * 0xdeadbeef + idx as u128),
                name: t.name.to_string(),
                size: t.size,
                progress: t.progress,
                state: t.state.clone(),
                download_speed: dl_speed,
                upload_speed: ul_speed,
                eta,
            });
        }

        SystemMetrics {
            id: self.system_id.clone(),
            name: self.name.clone(),
            system_type: SystemType::Qbittorrent,
            status: SystemStatus::Healthy,
            last_updated: Utc::now(),
            error: None,
            metrics: MetricsPayload::QBittorrent(QBittorrentMetrics {
                download_speed: total_download_speed,
                upload_speed: total_upload_speed,
                active_downloads,
                active_uploads: 0,
                total_torrents: torrent_infos.len() as i32,
                torrents: torrent_infos,
            }),
        }
    }

    async fn close(&mut self) {}
}
